src/hvm.cu

```diff
@@ -1802,16 +1827,229 @@ Port gnet_expand(GNet* gnet, Port port) {
 Port gnet_make_node(GNet* gnet, Tag tag, Port fst, Port snd) {
   Port ret;
   Port* d_ret;
-  cudaMalloc(&d_ret, sizeof(Port)); 
+  cudaMalloc(&d_ret, sizeof(Port));
   make_node<<<1,1>>>(gnet, tag, fst, snd, d_ret);
   cudaMemcpy(&ret, d_ret, sizeof(Port), cudaMemcpyDeviceToHost);
   cudaFree(d_ret);
   return ret;
 }
 
+// Readback
+// --------
+
+// Reads back a λ-Encoded constructor from device to host.
+// Encoding: λt ((((t TAG) arg0) arg1) ...)
+Ctr gnet_read_ctr(GNet* gnet, Port port) {
+  Ctr ctr;
+  ctr.tag = -1;
+  ctr.args_len = 0;
+
+  // Loads root lambda
+  Port lam_port = gnet_expand(gnet, port);
+  if (get_tag(lam_port) != CON) return ctr;
+  Pair lam_node = gnet_node_load(gnet, get_val(lam_port));
+
+  // Loads first application
+  Port app_port = gnet_expand(gnet, get_fst(lam_node));
+  if (get_tag(app_port) != CON) return ctr;
+  Pair app_node = gnet_node_load(gnet, get_val(app_port));
+
+  // Loads first argument (as the tag)
+  Port arg_port = gnet_expand(gnet, get_fst(app_node));
+  if (get_tag(arg_port) != NUM) return ctr;
+  ctr.tag = get_u24(get_val(arg_port));
+
+  // Loads remaining arguments
+  while (TRUE) {
+    app_port = gnet_expand(gnet, get_snd(app_node));
+    if (get_tag(app_port) != CON) break;
+    app_node = gnet_node_load(gnet, get_val(app_port));
+    arg_port = gnet_expand(gnet, get_fst(app_node));
+    ctr.args_buf[ctr.args_len++] = arg_port;
+  }
+
+  return ctr;
+}
+
+// Reads back a UTF-16 string.
```

Comment:
strings in hvm are actually just lists of unicode codepoints, we must update that